crates/astria-conductor/src/executor/queue.rs

```diff
@@ -0,0 +1,683 @@
+            self.remove_data_below_height(self.head_height);
+            Some(returned_soft_blocks)
+        } else {
+            None
+        }
+    }
+
+    /// Return all the blocks at the head height
+    fn pop_head_blocks(&mut self) -> Option<Vec<SequencerBlockSubset>> {
+        if let Some(head_blocks) = self.pending_blocks.get_mut(&self.head_height) {
+            let mut output_blocks: Vec<SequencerBlockSubset> = vec![];
+            let tmp_blocks = head_blocks.clone();
+            let mut blocks: Vec<&SequencerBlockSubset> = tmp_blocks.values().collect();
+            blocks.sort();
+            for block in blocks {
+                output_blocks.push(block.clone());
+            }
+            let most_recent_height = output_blocks[output_blocks.len() - 1].height();
+            self.head_height = most_recent_height.increment();
+            self.remove_data_below_height(self.head_height);
+            return Some(output_blocks);
+        }
+        None
+    }
+
+    /// Check to see if the block is already present in the queue
+    fn is_block_present(&mut self, block: &SequencerBlockSubset) -> bool {
+        let block_hash = block.block_hash();
+        let height = block.height();
+
+        // check if the block is already present in the pending blocks
+        if let Some(pending_blocks) = self.pending_blocks.get(&height) {
+            if let Some(_block) = pending_blocks.get(&block_hash) {
+                return true;
+            }
+        }
+        // check if the block is already present in the soft blocks
+        if let Some(soft_block) = self.soft_blocks.get(&height) {
+            if soft_block.block_hash() == block_hash {
+                return true;
+            }
+        }
+
+        false
+    }
+
+    /// Check if there is another block in the pending blocks at a lower height
+    /// that points to this block as its parent.
+    fn is_block_a_parent(&mut self, block: SequencerBlockSubset) -> bool {
+        let block_hash = block.block_hash();
+        if let Some(child_blocks) = self.pending_blocks.get(&block.child_height()) {
+            let blocks = child_blocks.values();
+            for block in blocks {
+                if let Some(hash) = block.parent_hash() {
+                    if hash == block_hash {
+                        return true;
+                    }
+                }
+            }
+        }
+        false
+    }
+
+    /// Safely insert a block into the pending blocks. This function will insert
+    /// the block into existing maps or create a new one if needed.
+    fn insert_to_pending_blocks(&mut self, block: SequencerBlockSubset) {
+        let height = block.height();
+        let block_hash = block.block_hash();
+
+        if let Some(pending_blocks) = self.pending_blocks.get_mut(&height) {
+            let mut new_map = pending_blocks.clone();
+            new_map.insert(block_hash, block);
+            self.pending_blocks.insert(height, new_map);
```

Comment:
remove and re-insert the map so you don't have to clone: https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#method.remove

Reply:
I am assuming this is being thought about in a thread safe way, and hence why atomically updating the entry for the `height` key in pending blocks is valuable. since hash map isn't thread safe by itself, I don't think it's necessary at this stage. I suggest avoiding the extra heap allocation with clone:
```suggestion
pending_blocks.insert(block_hash, block);
```